### PR TITLE
Refactor Collector class

### PR DESCRIPTION
**Pull request by intelliot**
## High Level Overview of Change

This refactors the Collector class to be aware of the new data models.

It adds a new Connection class which handles a single connection to one rippled server.

This replaces the ValidationStream and Network classes, which have been removed.

### Context of Change

Previously, the Network and ValidationStream classes stored everything in memory, and used these hash maps to keep track of which validators are currently in the UNL. The database is now the SSOT (Single Source Of Truth) for everything we know from the default validator list and from manifests received over the p2p network.

### Type of Change

- [x] Refactor (non-breaking change that only restructures code)

## Test Plan

Two new npm scripts have been added in package.json:

- connection-demo
- collector-demo

Use `yarn connection-demo` and `yarn collector-demo` to run these.

### CODE REVIEW

File: src/collector.ts [hunk 1]
```diff
@@ -1,503 +1,360 @@
...
- * - Deduplicates messages
- * - Stores validations
+ * Collector class
+ *
+ * Retrieves the UNL, connects to mulitple rippled servers, and filters out validations from non-UNL validators.
+ *
+ * TODO:
+ * - How much disk is required to store 24 hours worth of validations?
+ * - Create online_delete-like process that optionally deletes validations older than 24 hours
+ * - Deduplicate messages elsewhere
```
Comment (hbergren) [inline src/collector.ts]: Is there an issue that captures these TODOs?

File: src/collector.ts [hunk 2]
```diff
@@ -1,503 +1,360 @@
...
-  // The identifying hash of the proposed ledger is being validated.
-  ledger_hash: string // PK
-
-  // The Ledger Index of the proposed ledger.
-  ledger_index: string // Integer // PK
+interface ManifestData {
+  validatorName: string,
+  validation_public_key_base58: string,
+  signing_key: string,
+  manifestSequence: number,
```
Comment (hbergren) [inline src/collector.ts]: Is there a reason we're mixing snake_case and camelCase here?

File: src/collector.ts [hunk 3]
```diff
@@ -1,503 +1,360 @@
...
 
-// Associates a signing_key with a master_key (pubkey).
-export interface ManifestMessage {
+export interface ValidatorValidation extends ValidationMessage {
+  validator: Validator
+}
 
-  // The value `manifestReceived` indicates this is from the manifests stream.
-  type: 'manifestReceived'
+export type NetworkType = 'testnet' | 'mainnet'
```
Comment (hbergren) [inline src/collector.ts]: Misc: Is `export type NetworkType = 'testnet' | 'mainnet'` preferable to
```ts
export enum NetworkType {
  Testnet = 'testnet',
  Mainnet = 'mainnet',
}
```

I'm still trying to understand how to best handle fixed-cardinality types like these.

File: src/entity/Validator.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
-import {Entity, Column, PrimaryColumn, OneToMany} from "typeorm";
+import {Entity, Column, PrimaryColumn, OneToMany, UpdateDateColumn, Index} from "typeorm";
 import {Validation} from "./Validation";
 import {Manifest} from "./Manifest";
 
 @Entity()
+@Index(["signing_key", "validatorListSequence"], { unique: true })
```
Comment (hbergren) [inline src/entity/Validator.ts]: Is this a unique constraint, or an index for querying, or both?

What does this unique constraint mean in plain English?

File: src/connection.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+  // (May be omitted) The amendments this server wants to be added to the protocol.
+  amendments?: string[] // PK
+
+  // (May be omitted) The unscaled transaction cost (`reference_fee` value) this
+  // server wants to set by Fee Voting.
+  base_fee?: number // Integer // PK
+
+  // Bit-mask of flags added to this validation message.
+  // The flag 0x80000000 indicates that the validation signature is fully-canonical.
```
Comment (hbergren) [inline src/connection.ts]: What does "fully-canonical" mean?

File: src/connection.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  // The flag 0x00000001 indicates that this is a full validation; otherwise it's a partial validation.
+  // Partial validations are not meant to vote for any particular ledger.
+  // A partial validation indicates that the validator is still online but not keeping up with consensus.
+  flags: number // PK
+
+  // If true, this is a full validation. Otherwise, this is a partial validation.
+  // Partial validations are not meant to vote for any particular ledger.
+  // A partial validation indicates that the validator is still online but not keeping up with consensus.
+  full: boolean // PK
```
Comment (hbergren) [inline src/connection.ts]: Is this a computed property from flags?

File: src/connection.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  // When this validation vote was signed, in seconds since the Ripple Epoch.
+  signing_time: number // PK
+
+  // The base58 encoded public key from the key-pair that the validator used to sign the message.
+  // This identifies the validator sending the message and can also be used to verify the signature.
+  // Typically this is a signing key (signing_key). Use manifests to map this to an actual validation_public_key.
+  validation_public_key: string // PK
+
+  // The time when the validation message was received.
+  timestamp: moment.Moment
```
Comment (hbergren) [inline src/connection.ts]: `moment` is a pretty heavy library. Is it worth looking into alternatives? I'm not sure of the scope of `moment` functionality we're using in this repo right now.

File: src/connection.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+      }
+    })
+  }
+
+  // When you are finished with the WebSocket, it is fastest and most reliable to call
+  // terminate(). Using close() sometimes takes up to 10 seconds to complete.
+  terminate() {
+    this.ws.terminate()
+
+    // When closing or terminatingn the WebSocket connection, we must clear the
```
Comment (hbergren) [inline src/connection.ts]:
```suggestion
    // When closing or terminating the WebSocket connection, we must clear the
```

File: src/collector.ts (hunk 1 above)
Comment (intelliot) [reply src/collector.ts]: No, but as they are handled, they will be removed from the code.

File: src/collector.ts (hunk 2 above)
Comment (intelliot) [reply src/collector.ts]: snake_case comes from rippled JSON; camelCase is used for new names defined in this app. We can choose one and try to rename everything to match, but I don't think that's high priority right now.

File: src/collector.ts (hunk 3 above)
Comment (intelliot) [reply src/collector.ts]: I don't think it matters either way.

File: src/entity/Validator.ts (hunk 4 above)
Comment (intelliot) [reply src/entity/Validator.ts]: Both. No particular signing_key and validatorListSequence pair can occur more than once in the Validator table.

File: src/connection.ts (hunk 5 above)
Comment (intelliot) [reply src/connection.ts]: I think it's virtually the same as when the term is applied to transaction signatures: https://xrpl.org/transaction-malleability.html

File: src/connection.ts (hunk 6 above)
Comment (intelliot) [reply src/connection.ts]: I don't know, but it comes from rippled - so it's not computed by this app (or any TS/JS code).

File: src/connection.ts (hunk 7 above)
Comment (intelliot) [reply src/connection.ts]: I think we can probably remove this particular use of `moment`, but since this is a server-side app I don't think there's any significant cost in using `moment` either.

**Comment by intelliot** [comment]
Great comments, thanks!

Merging this since I have another giant PR that needs review...

(It makes a lot of changes to the code in this PR, so much of this will soon become obsolete.)
